### ISSUE TITLE
reload son file at runtime

### ISSUE THREAD

**Issue by mvanhelmont** [danielgerlag/workflow-core#110]
Thanks for the great framework, we currently experimenting with it. We are loading our son file add startup in our asp.net core application. This works fine and we can call workflows by our controllers.

The only problem that we had is when we change the file we need to restart de app as expected of course. Is there a way so we can refresh the LoadDefinition?

thanks,

**Comment by danielgerlag** [comment]
I believe you could just increment the version number in the JSON file, which would allow any already running workflows to continue on the old version and any new workflows the start subsequently, to run on the new version.

**Comment by germandb** [comment]
Hi mvanhelmont  in the RegisterWork method in the WorkflowRegistry.cs you would find 

`if (_registry.Any(x = x.Item1 == definition.Id && x.Item2 == definition.Version))
{
    throw new InvalidOperationException($"Workflow {definition.Id} version {definition.version} is already registered");
}`

You can change the throw exception and use a return instead, that way if the definition is already registered you dont do anything and in the other case the execution continue and registre the definition. I tried that and all the examples and tests run fine, but maybe Danielle can tell why the need of throw the exception.

**Comment by danielgerlag** [comment]
@mvanhelmont I'm not sure I'm following what you are doing... could you attach the code you are using to reload the workflow? as well as the 2 JSON versions if possible?

**Comment by mvanhelmont** [comment]
@danielgerlag 

I will try to explain. I what to use Workflow-core in my asp.net application. At the startup class i have this code:


```
            var host = app.ApplicationServices.GetService<IWorkflowHost>();
            var loader = app.ApplicationServices.GetService<IDefinitionLoader>();

            loader.LoadDefinition(System.IO.File.ReadAllText(env.WebRootPath + "/workflows.json"));

            host.Start();
```


this loads all workflows in memory. When i added a workflow to the json file i need to restart my application. So i have a controller with a action ReloadWorkflow. that uses IDefinitionLoader to load the json again. When i do this i get the message InvalidOperationException: Workflow HelloWorld version 3 is already registered.

So how do i load definitions add runtime when i create a new workflow or bump up the version number.

here are some json files:

```
{
  "Id": "HelloWorld",
  "Version": 1,
  "Steps": [
    {
      "Id": "Bye",
      "StepType": "WorkflowSample.Workflows.Steps.Step1, WorkflowSample",
    }
  ]
}
```

```
{
  "Id": "HelloWorld",
  "Version": 1,
  "Steps": [
    {
      "Id": "Bye",
      "StepType": "WorkflowSample.Workflows.Steps.Step1, WorkflowSample",
    }
  ],
  "Id": "HelloWorld",
  "Version": 2,
  "Steps": [
    {
      "Id": "Bye",
      "StepType": "WorkflowSample.Workflows.Steps.Step1, WorkflowSample",
      "NextStepId": "Bye2"
    },
    {
      "Id": "Bye2",
      "StepType": "WorkflowSample.Workflows.Steps.Step2, WorkflowSample"
    }
  ]
}
```

**Comment by danielgerlag** [comment]
Yes, that should work, and then load the newer definition of the only the changed workflow with the incremented version number. Let me you  if you get this to work.

**Comment by mvanhelmont** [comment]
@danielgerlag hi thanks i can confirm that this works so thats great. But the only problem i see when i use this approach is that loading all file on startup can be messy. especially if there a many workflows.

I have looked ate the IWorkflowRegistry implementation and maybe it's possible to replace the in memory store to a sql database perhaps?

I don't know what your thoughts of this approach is?

**Comment by danielgerlag** [comment]
I don't have any plans to do that, my approach would be to read your JSON definitions from a database at startup, rather than change the `WorkflowRegistry`

**Comment by pcopley** [comment]
@mvanhelmont You can load dynamically from JSON as well. You would need to test loading an identical workflow ID with an incremented version but I don't see why that would not work. Inject `IServiceRegistry` (and `IWorkflowController` if you're calling it immediately), then load the JSON.

    var loader = _serviceProvider.GetService<IDefinitionLoader>();
    var workflow = loader.LoadDefinition(yourWorkflowJsonWithIncrementedVersion);
    await _workflowController.StartWorkflow(workflow.Id);

**Comment by SSanthosh-SBT** [comment]
How to pass initial data by while using LoadDefinition??
